File: design/src/main/java/com/think/design/bottomnav/BottomNavigationFragment.java
/*
 * Copyright 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.think.design.bottomnav;

import com.think.design.R;

import androidx.fragment.app.Fragment;
import dagger.Provides;
import dagger.android.ContributesAndroidInjector;
import dagger.multibindings.IntoSet;
import com.think.design.application.scope.ActivityScope;
import com.think.design.application.scope.FragmentScope;
import com.think.design.feature.Demo;
import com.think.design.feature.DemoLandingFragment;
import com.think.design.feature.FeatureDemo;
import java.util.ArrayList;
import java.util.List;

/** A landing fragment that links to bottom nav demos for the Catalog app. */
public class BottomNavigationFragment extends DemoLandingFragment {

  @Override
  public int getTitleResId() {
    return R.string.cat_bottom_nav_title;
  }

  @Override
  public int getDescriptionResId() {
    return R.string.cat_bottom_nav_description;
  }

  @Override
  public Demo getMainDemo() {
    return new Demo() {
      @Override
      public Fragment createFragment() {
        return new BottomNavigationMainDemoFragment();
      }
    };
  }

  @Override
  public List<Demo> getAdditionalDemos() {
    List<Demo> additionalDemos = new ArrayList<>();
    additionalDemos.add(
        new Demo(R.string.cat_bottom_nav_label_visibility_demo_title) {
          @Override
          public Fragment createFragment() {
            return new BottomNavigationLabelVisibilityDemoFragment();
          }
        });
    additionalDemos.add(
        new Demo(R.string.cat_bottom_nav_animated_demo_title) {
          @Override
          public Fragment createFragment() {
            return new BottomNavigationAnimatedDemoFragment();
          }
        });
    return additionalDemos;
  }

  /** The Dagger module for {@link BottomNavigationFragment} dependencies. */
  @dagger.Module
  public abstract static class Module {
    @FragmentScope
    @ContributesAndroidInjector
    abstract BottomNavigationFragment contributeInjector();

    @IntoSet
    @Provides
    @ActivityScope
    static FeatureDemo provideFeatureDemo() {
      return new FeatureDemo(R.string.cat_bottom_nav_title, R.drawable.ic_bottomnavigation) {
        @Override
        public Fragment createFragment() {
          return new BottomNavigationFragment();
        }
      };
    }
  }
}

File: design/src/main/java/com/think/design/bottomnav/BottomNavigationFragment.kt
package com.think.design.bottomnav

import androidx.fragment.app.Fragment
import com.think.design.R
import com.think.design.application.scope.ActivityScope
import com.think.design.application.scope.FragmentScope
import com.think.design.feature.Demo
import com.think.design.feature.DemoLandingFragment
import com.think.design.feature.FeatureDemo
import dagger.Provides
import dagger.android.ContributesAndroidInjector
import dagger.multibindings.IntoSet

class BottomNavigationFragment : DemoLandingFragment() {
    override fun getTitleResId(): Int {
        return R.string.cat_bottom_nav_title
    }

    override fun getDescriptionResId(): Int {
        return R.string.cat_bottom_nav_description
    }

    override fun getMainDemo(): Demo {
        return object : Demo() {
            override fun createFragment(): Fragment? {
                return BottomNavigationMainDemoFragment()
            }
        }
    }

    override fun getAdditionalDemos(): MutableList<Demo?> {
        val additionalDemos: MutableList<Demo?> = ArrayList<Demo?>()
        additionalDemos.add(
            object : Demo(R.string.cat_bottom_nav_label_visibility_demo_title) {
                override fun createFragment(): Fragment? {
                    return BottomNavigationLabelVisibilityDemoFragment()
                }
            })
        additionalDemos.add(
            object : Demo(R.string.cat_bottom_nav_animated_demo_title) {
                override fun createFragment(): Fragment? {
                    return BottomNavigationAnimatedDemoFragment()
                }
            })
        return additionalDemos
    }

    /** The Dagger module for [BottomNavigationFragment] dependencies.  */
    @dagger.Module
    abstract class Module {
        @FragmentScope
        @ContributesAndroidInjector
        abstract fun contributeInjector(): BottomNavigationFragment?

        companion object {
            @JvmStatic
            @IntoSet
            @Provides
            @ActivityScope
            fun provideFeatureDemo(): FeatureDemo {
                return object :
                    FeatureDemo(R.string.cat_bottom_nav_title, R.drawable.ic_bottomnavigation) {
                    override fun createFragment(): Fragment {
                        return BottomNavigationFragment()
                    }
                }
            }
        }
    }
}
